# Question 5

(a) Implement your own code for a logistic regression classifier, which is trained
using gradient descent and cross-entropy error as the error function.

In [16]:
# importing required libraries

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from math import *

In [17]:
# Implementing own code for a logistic regression using gradient descent and cross entropy error function
def sigmoid(x,w):
    y = 1/(1+np.exp(-np.dot(x,w)))
    return y

def logistic_regression(X,Y):
    ## Data normalisation
    X = X-X.mean()
    ## intializing weights
    w = np.array([1,1])
    ## learning rate
    g = 0.01
    i=1
    while(1):
        y_pred = 1/(1+np.exp(-np.dot(w,X.T)))
        if(i==0):
            i+=1
            print("Weights after one iteration are: ",w)
        change = np.linalg.norm(g*(np.dot(X.T,Y-y_pred)))
        if(change>=1e-3):
            ## updating weights
            w = w+g*(np.dot(X.T,Y-y_pred))
        else:
            return w
        


(b)

In [28]:
def logistic_regression(X,Y):
    ## intializing weights
    w = np.array([-1,1.5,0.5])
    ## learning rate
    g = 0.1
    i=0
    while(1):
        y_pred = 1/(1+np.exp(-np.dot(w,X.T)))
        
        if(i==1):
            print("Weights after one iteration are: ",w)
        i+=1
        change = np.linalg.norm(g*(np.dot(X.T,Y-y_pred)))
        if(change>=1e-3):
            ## updating weights
            w = w+g*(np.dot(X.T,Y-y_pred))
        else:
            return w
X = np.array([[1,0.346,0.780],[1,0.303,0.439],[1,0.358,0.729],[1,0.602,0.863],[1,0.790,0.753],[1,0.611,0.965]])
Y = np.array([0,0,0,1,1,1])
w = logistic_regression(X,Y)
print("Final weights: ",w)


Weights after one iteration are:  [-1.01899756  1.53210518  0.51181202]
Final weights:  [-18.07774125  26.47820683   6.58909946]


(i) Final logistic model $P(\hat y=1|x_1,x_2)$ and its cross entropy function

$$\vec{w}=[-18.07774125  26.47820683   6.58909946]$$
$$P(\hat y=1|x_1,x_2)=\dfrac{1}{1+e^{-(-18.07774125+26.47820683x_1+6.58909946x_2)}}$$
$$P(\hat y=0|x_1,x_2)=1-\dfrac{1}{1+e^{-(-18.07774125+26.47820683x_1+6.58909946x_2)}}$$
$$J(\vec{w})=\displaystyle\sum_{i}y_ilog(P(\hat y_i=1|\vec{x}_i))+(1-y_i)log(P(\hat y_i=0|\vec{x}_i))$$
where
$$\vec{x}=[1,x_1,x_2]$$

(ii) Use gradient descent to update $\theta_0,\theta_1,\theta_2$ for one iteration. Write down
the updated logistic regression model.

After one iteration, weights:
$$\vec{w}=[-1.01899756,  1.53210518,  0.51181202]$$
Logistic regression model:
$$P(\hat y=1|x_1,x_2)=\dfrac{1}{1+e^{-(-1.01899756+1.53210518x_1+0.51181202x_2)}}$$
$$P(\hat y=0|x_1,x_2)=1-\dfrac{1}{1+e^{-(-1.01899756+1.53210518x_1+0.51181202x_2)}}$$
$$J(\vec{w})=\displaystyle\sum_{i}y_ilog(P(\hat y_i=1|\vec{x}_i))+(1-y_i)log(P(\hat y_i=0|\vec{x}_i))$$
where
$$\vec{x}=[1,x_1,x_2]$$

(iii) At convergence of gradient descent, use the model to make predictions
for all the samples in the test dataset. Calculate and report the accuracy, precision
and recall to evaluate this model.

In [29]:
X_test = np.array([[1,0.959,0.382],[1,0.750,0.306],[1,0.395,0.760],[1,0.823,0.764],[1,0.761,0.874],[1,0.844,0.435]])
y_pred = 1/(1+np.exp(-np.dot(w,X_test.T)))
y_test = np.array([0,0,0,1,1,1])
y_pred = np.round(y_pred).astype(int)
print("Prediction: ",y_pred)
print("Actual: ",y_test)

Prediction:  [1 1 0 1 1 1]
Actual:  [0 0 0 1 1 1]


In [31]:

#Reporting accuracy
print('Accuracy: ',format((y_test == y_pred).sum().astype(float)/len(y_pred)))

Accuracy:  0.6666666666666666


In [32]:
# Reporting precision

def precision(y,y_pred):
    n = len(y)
    tp = 0
    fp = 0
    for i in range(n):
        tp += (y_pred[i] == 1 and y_test[i] == 1)
        fp += (y_pred[i] == 1 and y_test[i] == 0)
    p = tp/(tp+fp)
    return p
print("Precision: ",precision(y_test,y_pred))

Precision:  0.6


In [33]:
# Reporting recall

def recall(y,y_pred):
    n = len(y)
    tp = 0
    fn = 0
    for i in range(n):
        tp += (y_pred[i] == 1 and y_test[i] == 1)
        fn += (y_pred[i] == 0 and y_test[i] == 1)
    p = tp/(tp+fn)
    return p
print("recall: ",recall(y_test,y_pred))

recall:  1.0
